dali/operators/imgcodec/decoder.h

```diff
@@ -0,0 +1,202 @@
+
+#ifndef DALI_OPERATORS_IMGCODEC_DECODER_H_
+#define DALI_OPERATORS_IMGCODEC_DECODER_H_
+
+namespace dali {
+namespace imgcodec {
+
+template <typename Backend>
+class DecoderBase : public Operator<Backend> {
+ public:
+  ~DecoderBase() override = default;
+
+ protected:
+  explicit DecoderBase(const OpSpec &spec) : Operator<Backend>(spec) {
+    output_format_ = spec.GetArgument<DALIImageType>("output_type");
+    dtype_ = spec.GetArgument<DALIDataType>("dtype");
+    GetDecoderSpecificArguments(spec);
+  }
+
+  virtual void SetupRoiGenerator(const OpSpec &spec, const workspace_t<Backend> &ws) {}
+
+  virtual ROI GetRoi(const OpSpec &spec, const workspace_t<Backend> &ws, std::size_t data_idx,
+                     TensorShape<> shape) {
+    return {};
+  }
+
+  template<typename T>
+  void GetDecoderSpecificArgument(const OpSpec &spec, const std::string &name,
+                                  std::function<bool(const T&)> validator = [](T x){return true;}) {
+    if (spec.HasArgument(name)) {
+      auto value = spec.GetArgument<T>(name);
+      if (!validator(value)) {
+        DALI_FAIL(make_string("Invalid value for decoder-specific parameter ", name));
+      }
+      decoder_params_[name] = value;
+    }
+  }
+
+  void GetDecoderSpecificArguments(const OpSpec &spec) {
+    GetDecoderSpecificArgument<uint64_t>(spec, "hybrid_huffman_threshold");
+    GetDecoderSpecificArgument<size_t>(spec, "device_memory_padding");
+    GetDecoderSpecificArgument<size_t>(spec, "host_memory_padding");
+    GetDecoderSpecificArgument<size_t>(spec, "device_memory_padding_jpeg2k");
+    GetDecoderSpecificArgument<size_t>(spec, "host_memory_padding_jpeg2k");
+    GetDecoderSpecificArgument<float>(spec, "hw_decoder_load");
+    GetDecoderSpecificArgument<size_t>(spec, "preallocate_width_hint");
+    GetDecoderSpecificArgument<size_t>(spec, "preallocate_height_hint");
+    GetDecoderSpecificArgument<bool>(spec, "memory_stats");
+    GetDecoderSpecificArgument<bool>(spec, "use_fast_idct");
+  }
+
+  bool CanInferOutputs() const override { return true; }
+
+  void SetupShapes(const OpSpec &spec, const workspace_t<Backend> &ws,
+                    std::vector<OutputDesc> &output_descs, ThreadPool& tp) {
+    output_descs.resize(1);
+    auto &input = ws.template Input<CPUBackend>(0);
+    int nsamples = input.num_samples();
+    rois_.resize(nsamples);
+    SetupRoiGenerator(spec, ws);
+    std::vector<TensorShape<>> shapes(nsamples);
+
+    for (size_t i = 0; i < shapes.size(); i++) {
+      tp.AddWork([i, &input, &shapes, &ws, &spec, this] (int tid) {
+        auto shape = PeekShape(input[i]);
+        ROI roi = GetRoi(spec, ws, i, shape);
+        shape[2] = NumberOfChannels(output_format_, shape[2]);
+        if (roi.use_roi()) {
+          auto roi_shape = roi.shape();
+          for (int d = 0; d < roi_shape.size(); d++) {
+            shape[d] = roi_shape[d];
+          }
+        }
+        shapes[i] = shape;
+        rois_[i] = roi;
+      }, 0);
+    }
```

Comment:
Why not just use `ImageDecoder::CalculateOutputShape`?

Reply:
Done